pkg/networkplugin-syncer/handlers/ovn/handler.go

```diff
@@ -0,0 +1,132 @@
+package ovn
+
+import (
+	"errors"
+	"sync"
+
+	goovn "github.com/ebay/go-ovn"
+	clientset "k8s.io/client-go/kubernetes"
+
+	submV1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/event"
+	"github.com/submariner-io/submariner/pkg/networkplugin-syncer/handlers/ovn/nbctl"
+)
+
+var WaitingForLocalEndpoint = errors.New("Waiting for the local endpoint details before we can " +
+	"setup any remote endpoint related information, this will be retried.")
+
+type SyncHandler struct {
+	event.HandlerBase
+	syncMutex        sync.Mutex
+	k8sClientset     clientset.Interface
+	nbctl            *nbctl.NbCtl
+	nbdb             goovn.Client
+	sbdb             goovn.Client
+	localEndpoint    *submV1.Endpoint
+	remoteEndpoints  map[string]*submV1.Endpoint
+	lastOvnGwChassis string
+}
+
+func (ovn *SyncHandler) GetName() string {
+	return "ovn-sync-handler"
+}
+
+func (ovn *SyncHandler) GetNetworkPlugin() string {
+	return "OVNKubernetes"
+}
+
+func NewSyncHandler(k8sClientset clientset.Interface) *SyncHandler {
+	return &SyncHandler{
+		remoteEndpoints: make(map[string]*submV1.Endpoint),
+		k8sClientset:    k8sClientset,
+	}
+}
+
+func (ovn *SyncHandler) Init() error {
+	if err := ovn.initClients(); err != nil {
+		return err
+	}
+
+	if err := ovn.ensureSubmarinerInfra(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ovn *SyncHandler) LocalEndpointCreated(endpoint *submV1.Endpoint) error {
+	ovn.syncMutex.Lock()
+	defer ovn.syncMutex.Unlock()
+
+	ovn.localEndpoint = endpoint
+
+	return ovn.updateGatewayNode()
+}
+
+func (ovn *SyncHandler) LocalEndpointUpdated(endpoint *submV1.Endpoint) error {
+	ovn.syncMutex.Lock()
+	defer ovn.syncMutex.Unlock()
+
+	ovn.localEndpoint = endpoint
+
+	return ovn.updateGatewayNode()
+}
+
+func (ovn *SyncHandler) LocalEndpointRemoved(endpoint *submV1.Endpoint) error {
+	if ovn.localEndpoint.Name == endpoint.Name {
+		ovn.localEndpoint = nil
+	}
+
+	return nil
+}
+
+func (ovn *SyncHandler) RemoteEndpointCreated(endpoint *submV1.Endpoint) error {
+	ovn.syncMutex.Lock()
+	defer ovn.syncMutex.Unlock()
+
+	ovn.remoteEndpoints[endpoint.Name] = endpoint
+
+	return ovn.updateRemoteEndpointsInfra()
+}
+
+func (ovn *SyncHandler) RemoteEndpointUpdated(endpoint *submV1.Endpoint) error {
+	ovn.syncMutex.Lock()
+	defer ovn.syncMutex.Unlock()
+
+	ovn.remoteEndpoints[endpoint.Name] = endpoint
```

Comment:
Just a thought: We are updating the remote endpoint info in our local cache before we any cleanup associated to the old remoteEndpoint. Just wanted to bring up if we are not leaving any stale resources in the switches/routers.

Reply:
No, we read the existing resources later, and we calculate the difference toAdd and toRemove for everything... so whatever is in our cache becomes a declarative state of what we want on the system.